nixos/modules/services/backup/bupstash.nix

```
@@ -0,0 +1,603 @@
+    name = "bupstash-${name}";
+    value = {
+      description = "Bupstash group job for ${name}.";
+      unitConfig = {
+        StopWhenUnneeded = true;
+      };
+    };
+  };
+
+  mkJobTimer = name: job: {
+    name = "bupstash-${name}";
+    value = {
+      wantedBy = [ "timers.target" ];
+      timerConfig = {
+        Unit = "bupstash-${name}.target";
+        OnCalendar = job.startAt;
+      };
+    };
+  };
+
+  mkJobUnits =
+    name: job:
+    {
+      "bupstash-${name}-prehook" = mkIf (job.preHook != "") {
+        description = "Hook that runs before the ${name} bupstash backup.";
+
+        partOf = [ "bupstash-${name}.target" ];
+        wantedBy = [ "bupstash-${name}.target" ];
+
+        script = job.preHook;
+
+        serviceConfig = {
+          # Prevents the unit to run multiple times before the postHook has finished
+          RemainAfterExit = true;
+          StateDirectory = "bupstash";
+          Type = "oneshot";
+          WorkingDirectory = "/var/lib/bupstash";
+        };
+      };
+
+      "bupstash-${name}-posthook" = mkIf (job.postHook != "" || job.preHook != "") {
+        description = "Hook that runs after the ${name} bupstash backup.";
+
+        partOf = [ "bupstash-${name}.target" ];
+        wantedBy = [ "bupstash-${name}.target" ];
+
+        # Require the preHook Unit, so that in case of failure it does not try to run
+        requires = optional (job.preHook != "") "bupstash-${name}-prehook.service";
+
+        script = job.postHook;
+
+        serviceConfig =
+          # Default configuration, used in all cases
+          {
+            StateDirectory = "bupstash";
+            Type = "oneshot";
+            WorkingDirectory = "/var/lib/bupstash";
+          }
+          # When postHook is empty, then this unit only exists to stop the prehook
+          // (optionalAttrs (job.postHook == "") {
+            ExecStart = "systemctl stop bupstash-${name}-prehook.service";
+            # Stopping the prehook returns the SIGTERM value, interpret it as a success
+            SuccessExitStatus = [ "SIGTERM" ];
+          })
+          # When both postHook and preHook exist, we need to stop the prehook after running the posthook
+          // (optionalAttrs (job.preHook != "" && job.postHook != "") {
+            ExecStopPost = "systemctl stop bupstash-${name}-prehook";
+          });
+      };
+    }
+    // (mapAttrs'
+      (
+        repo-name: repo:
+        nameValuePair "bupstash-${name}-${repo-name}" {
+          description = ''
+            Job for the ${name} bupstash backup, using ${repo.desc}.
+          '';
+
+          documentation = [
+            "man:bupstash-put(1)"
+            "https://bupstash.io/doc/man/bupstash-put.html"
+          ];
+
+          path = [ pkgs.openssh ];
+
+          after = optional (job.preHook != "") "bupstash-${name}-prehook.service";
+          before = optional (job.postHook != "" || job.preHook != "") "bupstash-${name}-posthook.service";
```

```suggestion
          after = optional (job.preHook != "") "bupstash-${name}-prehook.service";
          before = optional (job.postHook != "" || job.preHook != "") "bupstash-${name}-posthook.service";
 
          wants = [
            "bupstash-${name}-prehook.service"
            "bupstash-${name}-posthook.service"
          ];
```

In my tests, which admittedly only targeted a single repository, calling the main service for a repo would not result in hooks being started. Turning around the relationship like this does call the hooks.